specreduce/wavelength_calibration.py

```diff
@@ -0,0 +1,248 @@
+        return cls(...)
+
+    def _do_refinement(self, input_spectrum):
+        if self.refinement_method == 'gaussian':
+            window_width = self.refinement_kwargs.get('range')
+            window = SpectralRegion((self.pixel-window_width)*u.pix,
+                                    (self.pixel+window_width)*u.pix)
+
+            input_model = Gaussian1D(mean=self.pixel, stddev=3,
+                                     amplitude=self.input_spectrum.flux[self.pixel])
+
+            fitted_model = fit_lines(self.input_spectrum, input_model, window=window)
+            new_pixel = fitted_model.mean.value
+
+        elif self.refinement_method == 'min':
+            window_width = self.refinement_kwargs.get('range')
+            new_pixel = np.argmin(self.input_spectrum.flux[self.pixel-window_width:
+                                                           self.pixel+window_width+1])
+            new_pixel += self.pixel - window_width
+
+        elif self.refinement_method == 'max':
+            window_width = self.refinement_kwargs.get('range')
+            new_pixel = np.argmax(self.input_spectrum.flux[self.pixel-window_width:
+                                                           self.pixel+window_width+1])
+            new_pixel += self.pixel - window_width
+
+        else:
+            raise ValueError(f"Refinement method {self.refinement_method} is not implemented")
+
+        return new_pixel
+
+    def refine(self, input_spectrum=None, return_object=False):
+        # finds the center of the line according to refinement_method/kwargs and returns
+        # either the pixel value or a CalibrationLine object with pixel changed to the refined value
+        if self.refinement_method is None:
+            return self.pixel
+        input_spectrum = self.input_spectrum if input_spectrum is None else input_spectrum
+        refined_pixel = self._do_refinement(input_spectrum)
+        if return_object:
+            return CalibrationLine(input_spectrum, self.wavelength, refined_pixel,
+                                   refinement_method=self.refinement_method,
+                                   refinement_kwargs=self.refinement_kwargs)
+
+        return refined_pixel
+
+    @cached_property
+    def refined_pixel(self):
+        # returns the refined pixel for self.input_spectrum
+        return self.refine()
+
+    @property
+    def with_refined_pixel(self):
+        # returns a copy of this object, but with the pixel updated to the refined value
+        return self.refine(return_object=True)
+
+    def __str__(self):
+        return f"CalibrationLine: ({self.wavelength}, {self.pixel})"
+
+    def __repr__(self):
+        return f"CalibrationLine({self.wavelength}, {self.pixel})"
+
+
+class WavelengthCalibration1D():
+
+    def __init__(self, input_spectrum, lines, model=Linear1D(), spectral_unit=u.Angstrom,
+                 fitter=LMLSQFitter(calc_uncertainties=True),
```

Comment:
should we try to avoid the astropy warning by using a different default fitter, or maybe by detecting whether `model` is linear and choosing appropriately?

Reply:
Good call, I updated to check whether the model is linear or not and use a different fitter accordingly